PXB-3269 Reduce the time the Server is locked by xtrabackup

aybek opened the pull request.

altmannmarcelo committed:
PXB-3034 - Adding test cases

https://jira.percona.com/browse/PXB-3034

Added test cases under suite/lockless
Fixed test cases using --lock-ddl=false/true

altmannmarcelo committed:
PXB-3034 - Make --lock-ddl option an ENUM

https://jira.percona.com/browse/PXB-3034

Changed lock-ddl option to be an enum. Possible Values are:

ON - Same as True
OFF - Same as False
REDUCED - Enable REDUCED lock mode. The first attempt to copy IBD
files are done without locking and changed tables (Affected by DDL)
are recopied (if needed) under DDL.

altmannmarcelo committed:
PXB-3034 - Add DDL tracking to xtrabackup

https://jira.percona.com/browse/PXB-3034

Add DDL tracking to xtrabackup. This new object is responsible for
tracking DDL's while the backup is running.
Later those changes will be handled in the end of backup and during
prepare.

altmannmarcelo committed:
PXB-3034 - Handle prepare

https://jira.percona.com/browse/PXB-3034

Adjusted DDL tracking to produce correct files at the end of backup and
handle those files during prepare.

altmannmarcelo committed:
PXB-3034 - Second phase copy Multi thread

https://jira.percona.com/browse/PXB-3034

Added parallel copy capability to the second phase copy of .ibd files.

altmannmarcelo committed:
PXB-3034 - adjust fil_open_for_xtrabackup

Adjusted fil_open_for_xtrabackup to tolerate file been gone and
re-attempt to open the file 10 times.

aybek committed:
1. Use space_id instead of table name for naming .del .ren files.
2. PXB-3220 Allow deleted tables between disovery and file open, save them in missing tables list
3. PXB-3227: Rename table and then drop table, make sure that original table was deleted

aybek committed:
PXB-3238 handle rename ddls that happen during scan

aybek committed:
We want to note that table will be copied only after it has been opened and loaded to cache;
moving ddl_tracker->add_table to the correct spot

satya-bodapati committed:
PXB-3113 : Improve debug sync framework to allow PXB to pause and resume threads

https://perconadev.atlassian.net/browse/PXB-3113

The current debug-sync option in PXB completely suspends PXB process and user can resume by sending SIGCONT signal
This is useful for scenarios where PXB is paused and do certain operations on server and then resume PXB to complete.

But many bugs we found during testing, involves multiple threads in PXB. The goal of this work is to be able to
pause and resume the thread.

Since many tests use the existing debug-sync option, I dont want to disturb these tests. We can convert them to
the new mechanism later.

How to use?
-----------
The new mechanism is used with option --debug-sync-thread="sync_point_name"

In the code place a debug_sync_thread(“debug_point_1”) to stop thread at this place.

You can pass the debug_sync point via commandline --debug-sync-thread=”debug_sync_point1”

PXB will create a file of the debug_sync point name in the backup directory. It is suffixed with a threadnumber.
Please ensure that no two debug_sync points use same name (it doesn’t make sense to have two sync points with same name)

```
2024-03-28T15:58:23.310386-00:00 0 [Note] [MY-011825] [Xtrabackup] DEBUG_SYNC_THREAD: sleeping 1sec.  Resume this thread by deleting file /home/satya/WORK/pxb/bld/backup//xb_before_file_copy_4860396430306702017
```
In the test, after activating syncpoint, you can use wait_for_debug_sync_thread_point <syncpoint_name>

Do some stuff now. This thread is sleeping.

Once you are done, and if you want the thread to resume, you can do so by deleting the file 'rm backup_dir/sync_point_name_*`
Please use resume_debug_sync_thread_point <syncpoint_name> <backup_dir>. It dletes the syncpoint file and additionally checks that syncpoint is
indeed resumed.

More common/complicated scenario:
----------------------------------
The scenario is to signal another thread to stop after reaching the first sync point. To achieve this. Do steps 1 to 3 (above)

Echo the debug_sync point name into a file named “xb_debug_sync_thread”. Example:

4. echo "xtrabackup_copy_logfile_pause" > backup/xb_debug_sync_thread

5. send SIGUSR1 signal to PXB process. kill -SIGUSR1 496102

6. Wait for syncpoint to be reached. wait_for_debug_sync_thread <syncpoint_name>

PXB acknowledges it
2024-03-28T16:05:07.849926-00:00 0 [Note] [MY-011825] [Xtrabackup] SIGUSR1 received. Reading debug_sync point from xb_debug_sync_thread file in backup directory
2024-03-28T16:05:07.850004-00:00 0 [Note] [MY-011825] [Xtrabackup] DEBUG_SYNC_THREAD: Deleting  file/home/satya/WORK/pxb/bld/backup//xb_debug_sync_thread

and then prints this once the sync point is reached.
2024-03-28T16:05:08.508830-00:00 1 [Note] [MY-011825] [Xtrabackup] DEBUG_SYNC_THREAD: sleeping 1sec.  Resume this thread by deleting file /home/satya/WORK/pxb/bld/backup//xb_xtrabackup_copy_logfile_pause_10389933572825668634

At this point, we have two threads sleeping at two sync points. Either of them can be resumed by deleting the filenames mentioned in the error log.
(Or use resume_debug_sync_thread())

satya-bodapati committed:
PXB-3252 : Xtrabackup failed to read page after 10 retries. File ./mysql.ibd seems to be corrupted.

https://perconadev.atlassian.net/browse/PXB-3252

Problem:
--------
With lock-ddl=REDUCED, ALTER ENCRYPTION='Y'/'N' happens. On general tablespaces, this is done inplace.
ie the space_id of tablespace will not change and the pages are encrypted or decrypted.

For file per table tablespaces, a new tablespace is created with encryption key and data is copied from
old tablespace to new tablespace.

In xtrabackup, the files are discovered and then they are copied. Between these two operations, the encrypted
tablespace can change. For example, PXB saw that ts1.ibd is encrypted with key1, loaded into cache.

Then server did ENCRYPTION='N' and then back to ENCRYPTION='Y', now the tablspace is encrypted with a different key.

Now PXB copy threads tries to copy this tablespce and cannot decrypt a page. Page 0 is always unencrypted. So the
problem typically detected at Page 1. It can happen on any page.

Since PXB cannot decrypt the page, it reports corruption and aborts the backup.

Fix:
----
On decryption errors, we track such tablespaces with separate corrupted list. We also them to the recopy tables list.
Under lock, these tablespaces are copied again. A .new extension is used.
Then we process the corrupted list under lock. Create .corrupt files for the tablespaces from the corrupted list.
For example, if the tablespace encrypted is ts1.ibd, the file will be ts1.ibd.corrupted.

On prepare, we delete the corresponding ts1.ibd if the ts1.ibd.corrupted is present. This has to be done before the
*.ibd scan becuase tablespace loading aborts on processing such half-written tablespaces.
If the .corrupted is present in incremental directory, delete the ts1.ibd.meta and ts.ibd.delta files from the incremental
backup directory.

satya-bodapati committed:
PXB-3246 : Assertion failure: log0recv.cc:2141:!page || fil_page_type_is_index(page_type)

Problem:
--------
Unable to apply redo log record entry because page is in wrong state. It was observed that
tablespace is created by incremental backup

How did this happen?
--------------------

lets say tablespace is t1.ibd and happily in fullbackup
before incremental, this gets renamed to t2.ibd
incremental backup creates t2.ibd.delta and t2.ibd.meta files in incremental backup directory
later there is drop t2.ibd,  we have space_id.del file in incremental backup directory
also some redo generated on this table before it is dropped.

During prepare of incremental backup, when we process a space_id.del file, we check the tablespace if tablespace is found.
Lets say, it 2.del. To process 2.del, we first check, the tabespace that is with space_id 2.
Since the tablespace name is t1.ibd in the full backup directory, we delete it. Additionally,
we delete the .ibd and .meta files, so we try to delete t1.ibd.meta and t1.ibd.delta files.
They never existed, so we ignore the errors to delete them.

But in the inc backup directory, we still have t2.ibd.delta and t2.ibd.meta files. So inc backup prepare
creates a tablespace with space_id 2 and apply the delta file changes. This tablespace is wrong
because, we are creating a dropped tablespace and we dont have all the changes. incremental backup
creates this tablespace with all-zero 7 pages. Later when we do MLOG_INSERT into the index page,
we find out the page is NOT in correct state.

Fix:
----
We have to delete the right incremental files based on space_id. So we build metamap by scanning
*.meta files and with the key as space_id (found in meta file).

Later, when we process the space_id.del file, after removing the tablespace with space_id,
we will now ask aka meta map cache to give the .delta and .meta file belonging to deleted space_id.
By deleting the un-necessary .meta file and .delta, the tablespace is considred as dropped by redo
and corresponding redo entries are not applied.

satya-bodapati committed:
PXB-3253 : [ERROR] [MY-012592] [InnoDB] Operating system error number 2 in a file operation

https://perconadev.atlassian.net/browse/PXB-3253

Problem:
--------
Files disappear during backup with --lockd-ddl=reduced

Analysis:
---------
PXB open server files using os_file_create_simple_no_error_handling() via Fil_shard::open_file(),
Fil_shard::get_file_size(), Datafile::open_read_only. This API doesn't tolerate file open errors.

This particular bug occurs when the file disappeared after get_file_size() in Fil_shard::open_file().
(See the testcase for more details).

Fix:
----
If lock ddl is reduced and if we have not yet acquired/entered the copy under lock phase
ie is_server_locked() is false, we can tolerate the file open errors. So we use the function/API
os_file_create() instead of other variants. Within this, based on lock_ddl reduced mode, we
tolerate file opening errors.

satya-bodapati committed:
PXB-3223 : PXB must not allow --lock-ddl=REDUCED when pagetracking is enabled

Problem:
-------
We cannot allow pagetracking with lock-ddl=REDUCED. This is because page-tracking gives
us a set of page_ids (space_id, page_nos). PXB should copy these pages and while we copy
these pages, tablespace disappear, get renamed, encrypted etc.

We will enable it if there is need or usecase for this. For now, we will disable it.

Fix:
----
Disable the combination of --page-tracking and --lock-ddl=REDUCED

satya-bodapati committed:
PXB-3120 : Assertion failure: Dir_Walker::is_directory

Problem:
--------
InnoDB assumes directories or files do not disappear. It is true
for the engine because, it is in the startup mode and no opeartions are allowed
at this point of time.

Analysis:
---------
With lock-ddl=RECUCED, tables can be dropped concurrently when pxb does *.ibd scan
or subdirectories can disappear too.

Fix:
----
Handle walk_posix() for missing files/directories. The scan should continue and skip
these deleted files or directories.

satya-bodapati committed:
PXB-3281 : With lock-ddl=REDUCED, STL containers used by reduced code are not thread safe

Problem:
-------
xtrabackup uses multiple threads to scan the *.ibd files. With lock-ddl=reduced, we use several STL maps to track of missing, dropped or renamed tables.

Multiple threads are used only when number of IBDs are more than 8K

Unsafe calls:
  ddl_tracker->add_missing_table(phy_filename);
  ddl_tracker->add_renamed_table(space_id, path);

These calls from multiple threads operate on std::map/unordered_map and can cause race conditions.

Fix:
----
1. stream line mutex usage for entire ddl_tracker class. Currently used only for corrupted STL map.
2. Use space id instead of table id in messages
3. Rename add_table() since the name is confusing. Actual map elements should be renamed. it will be done later

satya-bodapati committed:
PXB-3241 : Assertion failure: os0file.cc:3416:!exists while taking backups with lock-ddl=REDUCED

Problem:
--------
Backups taken with lock-ddl=reduced, prepare failed to complete.

Analysis:
---------
When handling .ren files, the destination file name already exists and this causes assertion failure. See the below backup log

```
102: 2024-02-28T12:15:49.061631-00:00 1 [Note] [MY-011825] [Xtrabackup] DDL tracking : LSN: 73749548 create table ID: 788 Name: test/#sql-1fc79d_13#p#p3.ibd

423: 2024-02-28T12:15:50.121767-00:00 1 [Note] [MY-011825] [Xtrabackup] DDL tracking : LSN: 74312497 rename table ID: 425 From: test/tt_28_p#p#p3.ibd To: test/#sql2-1fc79d-13#p#p3.ibd

870: 2024-02-28T12:15:50.609015-00:00 2 [Note] [MY-011825] [Xtrabackup] Copying ./test/#sql-1fc79d_13#p#p3.ibd to /home/mohit.joshi/dbbackup_28_02_2024/full/test/#sql-1fc79d_13#p#p3.ibd

1337 2024-02-28T12:15:51.183699-00:00 1 [Note] [MY-011825] [Xtrabackup] DDL tracking : LSN: 74967007 rename table ID: 788 From: test/#sql-1fc79d_13#p#p3.ibd To: test/tt_28_p#p#p3.ibd

1491: 2024-02-28T12:15:51.398615-00:00 2 [Note] [MY-011825] [Xtrabackup] Copying ./test/tt_28_p#p#p3.ibd to /home/mohit.joshi/dbbackup_28_02_2024/full/test/tt_28_p#p#p3.ibd

2115:  2024-02-28T12:15:52.209645-00:00 1 [Note] [MY-011825] [Xtrabackup] DDL tracking : LSN: 75267178 delete table ID: 425 Name: test/#sql2-1fc79d-13#p#p3.ibd
```

Whats going on here?

Lets say we have partition  p3 with space id 425. This is being altered. So partition algorithm does this:

1. create new copy of space_id: 788 (#sql1).
2. rename the existing table 425 to some temp taname (#sql2)
3. we copied the new copy space_id 788 (#sql1) to backup.
4. we also copied the space_id 425 with original name (p3).
5. Later we saw a rename file for the copied tablespace 788. (788.ren created with destination name as tt_28_p#p#p3.ibd

The rename file for space_id 425 is skipped, because we knowthat it is dropped. So only a .del file. Final state of backup is:

===

788 in backup with name #sql1
425 in backup with name p3
788.ren file-> 788 From: test/#sql-1fc79d_13#p#p3.ibd To: test/tt_28_p#p#p3.ibd
425.drop file

===

Now prepare starts to process .ren files
it tries to rename 788  from #sql1 to p3. but p3 already exists..

Fix:
----
we skip rename and other operations if we know that tablespace is going to be dropped.
In the above example, we skipped 425.ren file.

So while preparing, we should handle the .del files first. Then we are applying all the consolidated operations in a way.
Then .ren can be processed

satya-bodapati committed:
PXB-3245 : Assertion failure: fil0fil.cc:2545:err == DB_SUCCESS found during incremental backup with lock-ddl=REDUCED

Problem:
--------
File deleted between PXB discovery and opening the file. This time at Fil_shard::create_node.
It insists the file to be found.

Fix:
----
1. Tolerate the file missing error
2. Use different error code to track in missing files
3. free the tablespace object on error (otherwise, if fil_space_t remians in cache, pxb will try to copy the file)

satya-bodapati committed:
PXB-3280 : undo log truncation causes assertion failure with reduced lock

Problem:
-------
With lock-ddl=reduced and Concurrent undo truncations, xtrabackup fails with an
assertion

Analysis:
---------
After truncation, the tablespace id of an undo tablespace id might change and xtrabackup returns error instead of crash. But higher layers of undo discovery do not
tolerate missing files or errors.

Fix:
---
Tolerate file deletions during undo discovery

satya-bodapati committed:
PXB-3034: Bring back UNIV_DEBUG on debug-sync-thread.

This will be debug only option. Fix release build issues

satya-bodapati committed:
PXB-3320 : prepare_handle_del_files() fails to delete the .meta and .delta files for deleted tablespaces in incremental backup directory

Problem:
--------
    1. take full backup with --lock-ddl=reduced

    2. create table t1(a INT), lets say space_id 10

    3. start incremental backup and pause before backup_start() function (we take Bakcup lock here)

    4. incremental backup copied t1.ibd.meta and t1.ibd.delta by this time

    5. DROP TABLE t1

    6. resume the incremental backup. 10.del file is created

    7. prepare the full backup with --apply-log-only

    8. prepare incremental backup

    incremental backup prepare first processes the .del files. before this all tabelspaces are loaded via .ibd scan

    since there is no t1.ibd in backup directory( it is only present as meta and delta file) in incremental backup directory, space_id with 10 is not in cache.

    Hence prepare_handle_del_files() will not delete the files related to space_id 10.
    We end up with orphan .ibd or .ibu files. Server ignore orphan .ibd
    But if the tablespace is undo tablespace, orphan .ibu are not ignored by server.
    Server discovers them via *.ibu scan. This can lead to assertion failures.

satya-bodapati committed:
PXB-3318 : prepare_handle_ren_files(): failed to handle .ren files

Problem:
-------
1. take full backup
2. create table t1 before incremental backup
3. Take incremental backup under gdb and pause at backup_start
4. now rename t1 to t2
5. let it finish
6. prepare full
7. prepare incremental

It happens because tables created between full and incremental are copied as .delta/*.meta files and not as IBD files.
prepare_handle_ren_files() relies on *.ibd scan but this cannot work as the *.delta files are not yet loaded to fil_cache.

Fix:
----
Use the meta_map generated from *.meta scan. Use the space_id from space_id.ren to identify the correct .meta and delta files.
Rename the matched .meta and .delta files to the destination name stored in the .ren file

satya-bodapati committed:
PXB-3295 : Undo tablespaces are not tracked properly with lock-ddl=REDUCED

Problem:
--------
1.
Undo tablespaces are not tracked properly. Since undo tablespaces are not opened via
fil_open_for_xtrabackup(), they are not tracked as 'copied'. This leads to wrong
decisions in handle_ddl_operations.

2.
When new undo tablespaces are created,
Server doesn't write a MLOG_FILE_CREATE record and so these are missed by the tracking
system.

Fix:
----
1. track undo tablespaces that xtrabackup copies without lock (before lock state)
2. After lock is taken, undo tablespces are discovered again (after lock state)

With this before and after states, we now determine undo files to be deleted,
undo files to be copied. Truncated undo tablespace use different tablespace id, so
old undo file is marked as deleted and new version of undo tablespace is copied

For example undo_001.ibu of space_id 10 is truncated, the filename remains same
but it space_id becomes 11. xtrabackup creates 11.ibu.del for undo tablespace to be deleted.
then undo_001.ibu.new with space_id 11 is copied under lock

satya-bodapati committed:
PXB-3295: fix testcase

satya-bodapati committed:
Follow up fix for PXB-3318: handle rename source and destination as same file

satya-bodapati committed:
PXB-3221 : Assertion failure: page0cur.cc:1177:ib::fatal triggered during prepare/or next server startup

Problem:
--------
If there are tables created in system tablespace and if ALTER ADD INDEX/DROP INDEX is executed before the
backup lock is taken, system tablespace could end up in corrupted state.

This is because this operation is not redologged and we are supposed to recopy the system tablespace files.
But we dont track system tablesapce, neither reopen and recopy them. Hence this issue.

Fix:
----
1. Track system tablespace in the list of tables tracked/backedup
2. Removing tracking for tables in system tablespace except of recopy. Other operations can be played via redolog
3. Reopen system tablespace and recopy them as ibdata1.new/ ibdata2.new

satya-bodapati committed:
PXB-3331 : Assertion failure: fil0fil.cc:6422:success

Problem:
-------
During prepare, for backups taken with lock-ddl=ON, we did *.ibd scan before
recovery.

This is allowed only for lock-ddl=REDUCED.

Fix:
----
During prepare, do *.ibd scan and processing of .new, .del, .ren, .corrupt files only if lock-ddl=REDUCED

aybek committed:
PXB-3278 : Wrong parsing of MLOG_FILE_ redo log records with lock-ddl…
…=REDUCED

Problem:
--------
ddl_tracker_t::backup_file_op assumes the required redo bytes are always present. see the assertion len < 6.
But it may happen we sometimes receive redo less than that. In such cases, we return nullptr and let the caller read more read and retry

Fix:
----
fil_tablespace_redo_create()/rename()/delete() variants handle this problem by returning nullptr and reading more redo and retry.
Moved ddl tracker calls to track after the validation is done.

aybek committed:
PXB-3248 Multiple files found for the same tablespace ID

    1. On prepare phase when handling ddl files some of the data files were not loaded to cache because of the first page validation therefore were left without applying ddls on them.
       To tolerate this issue we should open and load data files to cache without validation, to do so we are using fil_tablespace_open_for_recovery() function instead of fil_open_for_xtrabackup().
    2. Remove macro checks for debug_sync_thread() temporarily, for QA testing

satya-bodapati committed:
PXB-3349 : Fix compilation errors on some platforms

Errors are caused by usage of const itertator cend(). Replaced with
end().

satya-bodapati committed:
PXB-3350 : Display verbose DDL states maintained in reduced lock mode

Print all maps, sets, vector used for reduced lock feature

satya-bodapati committed:
PXB-3330 : Show the total time that xtrabackup locked the server

xtrabackup can lock server with three types of Backup locks
1. On upstream, LOCK INSTANCE FOR BACKUP
2. On PS, LOCK TABLES FOR BACKUP
3. On upstream if there are myisam tables or non-innodb tables, we also
   take FLUSH TABLES WITH READ LOCK (this is stronger than backup lock,
   DMLs allowd)

Display the total time server is locked by xtrabackup.

2024-08-16T12:06:06.602134+01:00 0 [Note] [MY-011825] [Xtrabackup] Total time Server is locked by LOCK INSTANCE FOR BACKUP is: 1.634 seconds
2024-08-16T12:06:06.602153+01:00 0 [Note] [MY-011825] [Xtrabackup] Total time Server is locked by FLUSH TABLES WITH READLOCK is: 134.165 milliseconds

satya-bodapati committed:
Rename testsuite from lockless to reducedlock

satya-bodapati committed:
PXB-3351 : Fix undo.sh random failure

Test has wrong assumption that a open transaction will block all undo
truncations. For the tablespaces that are marked as inactive, they are
not blocked by open transaction.

Modified it to check if truncated tablespace is recopied after
truncation.

satya-bodapati committed:
Fix memory leak seen on Jenkins.
This is caused by fil_space_read_name_and_filepath) which expects
caller to free the memory for the returned values.

Use scope_guard and simply the freeing logic. It is now auto-freed
on scope exit

satya-bodapati committed:
PXB-3352 : Fix jenkins test failures

Fix t/bug1461735.sh and t/ddl.sh test failures. These test use
lock-ddl=OFF and DDL during the backup. Although unsupported,
testcase exists.

Regression introduced from PXB-3246 fix. A part of fix which suppressing
warnings about missing files (they are tracked and dleted by DDL
handling at startup of prepare), introduced a problem that tablespaces
are not delted by processing of MLOG_FILE_DLETE redos (happens only if
backup is taken with lock-ddl=OFF).

Fixed the logic, for backups taken with lock-ddl=reduced,
track the deletions by updating the recv_sys deleted and missing maps.

satya-bodapati committed:
PXB-3352 : Fix jenkins test failures

Fix memory leak seen by undo.sh

On error path, memory allocated is not freed.

Commit:
PXB-3334 Code cleanups in reduced lock feature

made some minor code changes for better readability

Commit:
PXB-3334 Code cleanups in reduced lock feature

removed argument `prep_handle_ddls` from all using functions, using xtrabackup_prepare instead

satya-bodapati committed:
Refactor code. Move it reduced lock code to ddl_tracker files

it-percona-cla commented:
[![CLA assistant check](https://cla-assistant.percona.com:443/pull/badge/signed)](https://cla-assistant.percona.com:443/percona/percona-xtrabackup?pullRequest=1603) <br/>All committers have signed the CLA.

Commit:
PXB-3334 Code cleanups in reduced lock feature

Add error handling for prepare_handle_rename() and prepare_handle_del() operations and some minor code refactoring

satya-bodapati commented on storage/innobase/fil/fil0fil.cc:
```diff
@@ -1645,6 +1650,8 @@ class Fil_system {
   @return DB_SUCCESS if the open was successful */
   [[nodiscard]] dberr_t open_for_recovery(space_id_t space_id);
 
+  [[nodiscard]] dberr_t open_for_reduced(const std::string &path);
```
Add function comment in doxygen style

aybek replied:
AB

satya-bodapati commented on storage/innobase/fil/fil0fil.cc:
```diff
@@ -3371,6 +3398,10 @@ static bool fil_space_free(space_id_t space_id, bool x_latched) {
   return true;
 }
 
+space_id_t fil_get_tablespace_id(const std::string &filename) {
+  return Fil_system::get_tablespace_id(filename);
+}
```
entire function under ifdef XTRABACKUP?

aybek replied:
AB

satya-bodapati commented on storage/innobase/fil/fil0fil.cc:
```diff
@@ -10183,6 +10208,22 @@ dberr_t fil_tablespace_open_for_recovery(space_id_t space_id) {
+
+  auto status = ibd_open_for_recovery(space_id, path, space);
+
+  if (status == FIL_LOAD_DBWLR_CORRUPTION) {
+    return DB_CORRUPTION;
+  }
+  return DB_SUCCESS;
```
status is a enum and other states are not handled, please handle all of them and convert to DB_ * error codes.

satya-bodapati replied:
if we use fil_tablespace_open_for_recovery(), then this comment might be irrelevant?

aybek replied:
AB

satya-bodapati commented on storage/innobase/include/fil0fil.h:
```diff
@@ -2247,16 +2243,19 @@ void fil_set_scan_dir(const std::string &directory, bool is_undo_dir = false);
 @return DB_SUCCESS if open was successful */
 [[nodiscard]] dberr_t fil_tablespace_open_for_recovery(space_id_t space_id);
 
+dberr_t fil_open_for_reduced(const std::string &path);
```
Add function comment? Please add some comment this is a wrapper to fil_tablespace_open_for_recovery?

aybek replied:
AB

satya-bodapati commented on storage/innobase/fil/fil0fil.cc:
```diff
@@ -10183,6 +10208,22 @@ dberr_t fil_tablespace_open_for_recovery(space_id_t space_id) {
+  space_id_t space_id = get_tablespace_id(path);
+  fil_space_t *space;
+
+  auto status = ibd_open_for_recovery(space_id, path, space);
```
once we get space_id, can we just call fil_tablespace_open_for_recovery(space_id) ?

aybek replied:
if we do that Fil_system::open_for_recovery() will fail with assertion "ut_ad(recv_recovery_is_on() || Log_DDL::is_in_recovery());"

satya-bodapati replied:
The assertion should be handled, or lock_ddl == reduced

satya-bodapati replied:
Looks like some other problems exist. Dropping this comment

satya-bodapati commented on storage/innobase/fil/fil0fil.cc:
```diff
@@ -10505,11 +10546,16 @@ static void fil_make_abs_file_path(const char *file_name, ulint flags,
-                                       ulint parsed_bytes, bool parse_only) {
+const byte *fil_tablespace_redo_create(
+    const byte *ptr, const byte *end, const page_id_t &page_id,
+    ulint parsed_bytes, bool parse_only IF_XB(, lsn_t start_lsn)) {
```
lets rename this to record_lsn instead of start_lsn.  This function and other fil_tablespace_redo_* functions.

aybek replied:
AB

satya-bodapati commented on storage/innobase/fil/fil0fil.cc:
```diff
@@ -11543,15 +11644,37 @@ void Tablespace_dirs::open_ibd(const Const_iter &start, const Const_iter &end,
+        cache and rename ddl will be handled by ddl_tracker we skip loading
+        duplicate tablespace */
+        continue;
+      } else
+        /* PXB-2275 - Allow DB_INVALID_ENCRYPTION_META as we will test it in
+        the end of the backup */
+        if (err != DB_SUCCESS && err != DB_INVALID_ENCRYPTION_META) {
+          result = false;
+        }
```
can you please put braces aroudn the the last else {}

aybek replied:
AB

satya-bodapati commented on storage/innobase/fsp/fsp0file.cc:
```diff
@@ -122,9 +122,11 @@ dberr_t Datafile::open_read_only(bool strict) {
-  m_handle = os_file_create_simple_no_error_handling(
-      innodb_data_file_key, m_filepath, m_open_flags, OS_FILE_READ_ONLY, true,
-      &success);
+
+  m_handle = os_file_create(
+      innodb_data_file_key, m_filepath,
+      OS_FILE_OPEN | OS_FILE_ON_ERROR_SILENT | OS_FILE_ON_ERROR_NO_EXIT,
+      OS_FILE_NORMAL, OS_DATA_FILE, OS_FILE_READ_ONLY, &success);
```
I think, I mistakenly removed upstream code. Better to put xtrbackup code under #ifdef XTRABACKUP.

THe reason for this change is 
/* File discovery under xtrabackup should tolerate disappearing IBD files on server and we don't want to log warnings or errors. The DDL  tracker will track these deletions */

aybek replied:
Do you mean we should readd some code that was mistakenly removed and put it under #ifdef XTRABACKUP ? 
Not sure which part of the code you mean

satya-bodapati replied:
I just meant, using 
#ifdef XTRBACKUP
  xb_variant
#else
 upstream_variant
#endif

satya-bodapati commented on storage/innobase/include/fil0fil.h:
```diff
@@ -2111,27 +2110,25 @@ inline fil_space_t *fil_space_get_sys_space() {
 @param[in]      page_id         Tablespace Id and first page in file
 @param[in]      parsed_bytes    Number of bytes parsed so far
 @param[in]      parse_only      Don't apply, parse only
+@param[in]      start_lsn       LSN of the redo record
```
as noted earlier, lets rename this to record_lsn

aybek replied:
AB

satya-bodapati commented on storage/innobase/include/fil0fil.h:
```diff
@@ -2375,4 +2380,11 @@ inline bool fil_node_t::is_offset_valid(os_offset_t byte_offset) const {
 }
 
+/** Frees a space object from the tablespace memory cache.
+Closes a tablespaces' files but does not delete them.
```
remove the extra '

aybek replied:
AB

satya-bodapati commented on storage/innobase/log/log0recv.cc:
```diff
@@ -1688,24 +1689,32 @@ static const byte *recv_parse_or_apply_log_rec_body(
+                      recv_sys->recovered_lsn + parsed_bytes <
+                          backup_redo_log_flushed_lsn ||
+                      opt_lock_ddl == LOCK_DDL_REDUCED),
+          start_lsn);
```
Lets not change the start_lsn to record_lsn here and the callers of this function. These are coming from upstream.
Only fil_tablespace_redo_*() or below is sufficient for us

aybek replied:
AB

satya-bodapati commented on storage/innobase/xtrabackup/src/backup_mysql.cc:
```diff
@@ -143,6 +143,16 @@ static int original_innodb_buffer_pool_dump_pct;
 static bool innodb_buffer_pool_dump;
 static bool innodb_buffer_pool_dump_pct;
 
+/* This could be either LOCK TABLES FOR BACKUP or LOCK INSTANCE */
+static xtrabackup::utils::time backup_lock_start_time =
+    xtrabackup::utils::INVALID_TIME;
+static xtrabackup::utils::time backup_lock_end_time =
+    xtrabackup::utils::INVALID_TIME;
+
+static xtrabackup::utils::time ftwrl_start_time =
+    xtrabackup::utils::INVALID_TIME;
+static xtrabackup::utils::time ftwrl_end_time = xtrabackup::utils::INVALID_TIME;
```
under ifdef XTRABACUP

aybek replied:
AB

satya-bodapati commented on storage/innobase/xtrabackup/src/ddl_tracker.cc:
```diff
@@ -0,0 +1,1041 @@
+                                            std::string name) {
+  Fil_path::normalize(name);
+  if (Fil_path::has_prefix(name, Fil_path::DOT_SLASH)) {
+    name.erase(0, 2);
```
instead of 2, strlen(Fil_path::DOT_SLASH) or DOT_SLASH.length()?

aybek replied:
AB

satya-bodapati commented on storage/innobase/xtrabackup/src/ddl_tracker.h:
```diff
@@ -0,0 +1,194 @@
+  std::tuple<filevec, filevec> handle_undo_ddls();
+
+ public:
+  void add_undo_tablespace(const space_id_t space_id, std::string name);
```
We have not clearly explained how undo tablespaces are tracked. We should add detailed function comment here.

aybek replied:
I think you can explain this better than me 😃

satya-bodapati replied:
Yes, I will address this.

satya-bodapati commented on storage/innobase/xtrabackup/src/ddl_tracker.cc:
```diff
@@ -0,0 +1,1041 @@
+
+void ddl_tracker_t::add_to_recopy_tables(space_id_t space_id, lsn_t start_lsn,
+                                         const std::string operation) {
+  // There should never be MLOG_INDEX_LOAD. However MLOG_WRITE_STRING on
```
There should never be MLOG_INDEX_LOAD on undo tablespaces.

aybek replied:
AB

satya-bodapati commented on storage/innobase/xtrabackup/src/ddl_tracker.cc:
```diff
@@ -0,0 +1,1041 @@
+  space_id_to_name_t *new_tables;
+} copy_thread_ctxt_t;
+
+static void data_copy_thread_func(copy_thread_ctxt_t *ctxt) {
```
I think these are copy of original copy thread code. Lets rename to copy_for_reduced().

aybek replied:
AB

satya-bodapati commented on storage/innobase/xtrabackup/src/ddl_tracker.cc:
```diff
@@ -0,0 +1,1041 @@
+
+  /* create THD to get thread number in the error log */
+  THD *thd = create_thd(false, false, true, 0, 0);
+  debug_sync_point("data_copy_thread_func");
```
debug_sync_point here should be renamed and the corresponding usage in testcase should be renamed too

aybek replied:
AB

satya-bodapati commented on storage/innobase/xtrabackup/src/ddl_tracker.cc:
```diff
@@ -0,0 +1,1041 @@
+  my_thread_end();
+}
+
+/**
```
Function comment first line missing.
Please give an example in the comment

aybek replied:
AB

satya-bodapati commented on storage/innobase/xtrabackup/src/ddl_tracker.cc:
```diff
@@ -0,0 +1,1041 @@
+  std::ostringstream str;
+  str << "Before UNDO files" << to_string(before_lock_undo) << "\n"
+      << "After UNDO files" << to_string(after_lock_undo) << "\n";
+  fprintf(stderr, "%s\n", str.str().c_str());
```
Add a comment here that xb::info() is not used intentionally, as there is a loss of newlines characters from the stream/string.

aybek replied:
AB

satya-bodapati commented on storage/innobase/xtrabackup/src/ddl_tracker.cc:
```diff
@@ -0,0 +1,1041 @@
+      << "Corrupted_tablespaces: " << to_string(corrupted_tablespaces) << "\n"
+      << "Missing tables: " << to_string(missing_tables) << "\n"
+      << "Renamed_during_scan: " << to_string(renamed_during_scan) << "\n";
+  fprintf(stderr, "%s\n", str.str().c_str());
```
same as above. comment on why fprintf() instead of xb::info().

aybek replied:
AB

satya-bodapati commented on storage/innobase/xtrabackup/src/ddl_tracker.cc:
```diff
@@ -0,0 +1,1041 @@
+      }
+
+    } else {
+      std::tie(err, std::ignore) = fil_open_for_xtrabackup(
```
what to do if we cannot open a tablespace after taking the lock ?
We should abort the backup here. May an exit() would do 
[ I prefer to return to the callers way up and cleanup all subsystem but that is big refactoring change. Lets ignore for now]

satya-bodapati commented on storage/innobase/xtrabackup/src/ddl_tracker.cc:
```diff
@@ -0,0 +1,1041 @@
+
+    } else {
+      std::tie(err, std::ignore) = fil_open_for_xtrabackup(
+          table->second, table->second.substr(0, table->second.length() - 4));
```
why 4?

aybek replied:
AB

satya-bodapati commented on storage/innobase/xtrabackup/src/ddl_tracker.cc:
```diff
@@ -0,0 +1,1041 @@
+  std::string new_space_name = name;
+  Fil_path::normalize(new_space_name);
+  if (Fil_path::has_prefix(new_space_name, Fil_path::DOT_SLASH)) {
+    new_space_name.erase(0, 2);
```
dont use 2, strlen() or length

aybek replied:
AB

satya-bodapati commented on storage/innobase/xtrabackup/src/ddl_tracker.cc:
```diff
@@ -0,0 +1,1041 @@
+  std::string new_space_name{name};
+  Fil_path::normalize(new_space_name);
+  if (Fil_path::has_prefix(new_space_name, Fil_path::DOT_SLASH)) {
+    new_space_name.erase(0, 2);
```
here and everywhere, about 2

aybek replied:
AB

satya-bodapati commented on storage/innobase/xtrabackup/src/ddl_tracker.cc:
```diff
@@ -0,0 +1,1041 @@
+        .start();
+  }
+
+  /* Wait for threads to exit */
+  while (1) {
+    std::this_thread::sleep_for(std::chrono::seconds(1));
+    mutex_enter(&count_mutex);
+    if (count == 0) {
+      mutex_exit(&count_mutex);
+      break;
+    }
+    mutex_exit(&count_mutex);
+  }
```
Print a msg after sleep? Waiting for the recopy of tables to be finished. Expecting count be 0, current count is ?

aybek replied:
AB

satya-bodapati commented on storage/innobase/xtrabackup/src/ddl_tracker.cc:
```diff
@@ -0,0 +1,1041 @@
+    mutex_exit(&count_mutex);
+  }
+
+  if (data_copying_error) {
```
Add xb::error() here before exit(). That there is problem with recopy tablespaces under lock (reduced mode)

aybek replied:
AB

satya-bodapati commented on storage/innobase/xtrabackup/src/ddl_tracker.cc:
```diff
@@ -0,0 +1,1041 @@
+  return {exists, exists ? it->second : ""};
+}
+
+/** Read file content into STL string */
```
modernize to c++17 (chatgpt version. Please use xb::info() instead cerr etc.)

```
#include <fstream>
#include <string>
#include <filesystem>
#include <stdexcept>

static std::string read_file_as_string(const std::string& file) {
    try {
        // Check if the file exists using std::filesystem
        if (!std::filesystem::exists(file)) {
            return "";
        }

        std::ifstream f(file, std::ios::in | std::ios::binary);
        if (!f) {
            return "";
        }

        // Read entire file into a string
        std::string content((std::istreambuf_iterator<char>(f)),
                            std::istreambuf_iterator<char>());

        return content;
    } catch (const std::filesystem::filesystem_error& e) {
        // Handle filesystem-related exceptions
        // (e.g., file doesn't exist, permissions issues)
        std::cerr << "Filesystem error: " << e.what() << '\n';
        return "";
    } catch (const std::ifstream::failure& e) {
        // Handle ifstream-related exceptions
        // (e.g., file couldn't be opened, read errors)
        std::cerr << "File read error: " << e.what() << '\n';
        return "";
    } catch (const std::exception& e) {
        // Catch all other standard exceptions
        std::cerr << "Error: " << e.what() << '\n';
        return "";
    }
}

```

aybek replied:
AB

satya-bodapati commented on storage/innobase/xtrabackup/src/ddl_tracker.cc:
```diff
@@ -0,0 +1,1041 @@
+  }
+}
+
+static void rename_file(const std::string &from, const std::string &to) {
+  if (my_rename(from.c_str(), to.c_str(), MY_WME)) {
+    xb::error() << "Can't rename " << from << " to " << to << " errno "
+                << errno;
+    exit(EXIT_FAILURE);
+  }
+}
+
+static void rename_force(const std::string &from, const std::string &to) {
+  MY_STAT stat_info;
+  /* check if dest folder exists */
+  std::string dest_dir = to.substr(0, to.find_last_of("/"));
+  /* create target dir if not exist */
+  if (!my_stat(dest_dir.c_str(), &stat_info, MYF(0)) &&
+      (my_mkdir(dest_dir.c_str(), 0750, MYF(0)) < 0)) {
+    xb::error() << "cannot mkdir: " << my_errno() << " " << dest_dir;
+    exit(EXIT_FAILURE);
+  }
+  rename_file(from, to);
+}
+
```
as noted in slack DM:

```
if source file doesn't exist but destination file exists with the desired name -> not an error
if source file doesn't exists but destination file DO NOT exist with desired name -> error
if source file exists, and desitnation file exists with desired name -> error (to be validated)
if source file exists, and desitnation file doesn't exist with desired name -> not an error
```

aybek replied:
AB

aybek replied:
I have added rename error handling logic in prepare_handle_ren_files() func. 
Here https://github.com/percona/percona-xtrabackup/pull/1603/files/65300ae9ee89262ea5de2e2f2ebff8737bc74f03#diff-0647c6584a748ffd0ebb9e7b4957e7ddbdcdfca51932a0c38c5960e5d7ec001eR825

satya-bodapati commented on storage/innobase/xtrabackup/src/ddl_tracker.cc:
```diff
@@ -0,0 +1,1041 @@
+  return (true);
+}
+
+/**
+ * Handle DDL for renamed files
+ * example input: test/10.ren file with content = test/new_name.ibd ;
+ *        result: tablespace with space_id=10 will be renamed to
+ * test/new_name.ibd
+ * @return true on success
+ */
```
parameter comment missing.

aybek replied:
AB

satya-bodapati commented on storage/innobase/xtrabackup/src/ddl_tracker.cc:
```diff
@@ -0,0 +1,1041 @@
+ * @return true on success
+ */
+bool prepare_handle_ren_files(
+    const datadir_entry_t &entry, /*!<in: datadir entry */
+    void * /*data*/) {
```
remove trailing comments, this is old style.. most likely this is copy paste

Applies to other places with parameter comments like this.

aybek replied:
AB

satya-bodapati commented on storage/innobase/xtrabackup/src/ddl_tracker.cc:
```diff
@@ -0,0 +1,1041 @@
+}
+
+/**
+ * Handle .corrupt files. These files should be removed before we do *.ibd scan
+ * @return true on success
+ * */
```
parameter comment fix

aybek replied:
AB

satya-bodapati commented on storage/innobase/xtrabackup/src/ddl_tracker.cc:
```diff
@@ -0,0 +1,1041 @@
+  if (xtrabackup_incremental) {
+    std::string delta_file = source_path + EXT_DELTA;
+    xb::info() << "prepare_handle_corrupt_files: deleting  " << delta_file;
+    os_file_delete_if_exists_func(delta_file.c_str(), nullptr);
+
+    std::string meta_file = source_path + EXT_META;
+    xb::info() << "prepare_handle_corrupt_files: deleting  " << meta_file;
+    os_file_delete_if_exists_func(meta_file.c_str(), nullptr);
+  } else {
+    xb::info() << "prepare_handle_corrupt_files: deleting  " << source_path;
+    os_file_delete_if_exists_func(source_path.c_str(), nullptr);
+  }
+
+  // delete the .corrupt file, we don't need it anymore
+  os_file_delete_if_exists_func(corrupt_path.c_str(), nullptr);
```
what about return types of os_file_delte_if_exists..()

aybek replied:
AB

satya-bodapati commented on storage/innobase/xtrabackup/src/ddl_tracker.h:
```diff
@@ -0,0 +1,194 @@
+  space_id_to_name_t drops;
+  /* For DDL operation found in redo log,  */
+  std::unordered_map<space_id_t, std::pair<std::string, std::string>> renames;
+  /** Tables that have been deleted between discovery and file open */
+  std::unordered_set<std::string> missing_tables;
```
instead of missing_tables, please use missing_afer_discovery.

aybek replied:
AB

satya-bodapati commented on storage/innobase/xtrabackup/src/ddl_tracker.h:
```diff
@@ -0,0 +1,194 @@
+  /** Check if table is in missing list
+  @param[in]  name  tablespace name */
+  bool is_missing_table(const std::string &name);
+  std::string convert_file_name(space_id_t space_id, std::string file_name,
+                                std::string ext);
```
missing function comment

aybek replied:
AB

satya-bodapati commented on storage/innobase/xtrabackup/src/ddl_tracker.h:
```diff
@@ -0,0 +1,194 @@
+bool xtrabackup_scan_meta(const datadir_entry_t &entry, /*!<in: datadir entry */
+                          void * /*data*/);
+
+bool prepare_handle_new_files(
+    const datadir_entry_t &entry, /*!<in: datadir entry */
+    void *arg __attribute__((unused)));
```
missing function comment

aybek replied:
AB

satya-bodapati commented on storage/innobase/xtrabackup/src/ddl_tracker.h:
```diff
@@ -0,0 +1,194 @@
+else return <false, ""> if it doesnt exist */
+std::tuple<bool, std::string> is_in_meta_map(space_id_t space_id);
+
+/**
+ * Handle DDL for renamed files
+ * example input: test/10.ren file with content = test/new_name.ibd ;
+ *        result: tablespace with space_id=10 will be renamed to
+ * test/new_name.ibd
+ * @return true on success
+ */
+bool prepare_handle_ren_files(
+    const datadir_entry_t &entry, /*!<in: datadir entry */
+    void * /*data*/);
+
+/**
+ * Handle .corrupt files. These files should be removed before we do *.ibd scan
+ * @return true on success
+ * */
+bool prepare_handle_corrupt_files(
+    const datadir_entry_t &entry, /*!<in: datadir entry */
+    void * /*data*/);
+
+/**
+ * Handle DDL for deleted files
+ * example input: test/10.del file
+ *        result: tablespace with space_id=10 will be deleted
+ * @return true on success
+ */
+bool prepare_handle_del_files(
+    const datadir_entry_t &entry, /*!<in: datadir entry */
+    void *arg __attribute__((unused)));
+
```
improper function comments, parameter comments.

aybek replied:
AB

satya-bodapati commented on storage/innobase/xtrabackup/src/xtrabackup.cc:
```diff
@@ -1558,6 +1576,18 @@ Disable with --skip-innodb-checksums.",
      (G_PTR *)&xtrabackup_debug_sync, (G_PTR *)&xtrabackup_debug_sync, 0,
      GET_STR, REQUIRED_ARG, 0, 0, 0, 0, 0, 0},
+
+#ifdef UNIV_DEBUG
```
lets use ifndef NDEBUG (same as below)

aybek replied:
AB

aybek replied:
AB

satya-bodapati commented on storage/innobase/xtrabackup/src/xtrabackup.cc:
```diff
@@ -2514,6 +2636,22 @@ static bool xtrabackup_read_info(char *filename) {
   } else if (xb_server_version < 80029) {
     cfg_version = IB_EXPORT_CFG_VERSION_V6;
   }
+  /* skip start_time, end_time, lock_time, binlog_pos, innodb_from_lsn,
+   * innodb_to_lsn, partial, incremental, format, compressed, encrypt */
+  for (int i = 0; i < 12; i++) {
```
I wish a better way of skipping parameter without the hardcoded 12. Not very important. Just a note.

satya-bodapati commented on storage/innobase/xtrabackup/src/xtrabackup.cc:
```diff
@@ -2752,6 +2890,30 @@ static bool xtrabackup_write_info(const char *filepath) {
+  }
+}
+
+lock_ddl_type_t ddl_lock_type_from_str(std::string type) {
+  if (type == "ON") {
+    return LOCK_DDL_ON;
+  } else if (type == "OFF") {
+    return LOCK_DDL_OFF;
+  } else if (type == "REDUCED") {
+    return LOCK_DDL_REDUCED;
+  } else {
+    ut_error;
+  }
+}
```
what if there is TRUE and FALSE. This is allowed too. Please update this.
What about case comparison?

aybek replied:
AB
I think its better to leave if comparisons because switch type is string

satya-bodapati commented on storage/innobase/xtrabackup/src/xtrabackup.cc:
```diff
@@ -2998,6 +3160,11 @@ const char *xb_get_copy_action(const char *dflt) {
 /* TODO: We may tune the behavior (e.g. by fil_aio)*/
 
 static bool xtrabackup_copy_datafile(fil_node_t *node, uint thread_n) {
+  return xtrabackup_copy_datafile(node, thread_n, nullptr);
```
lets call the overloaded function by idfferent name. may be add _func or _low().

aybek replied:
AB

satya-bodapati commented on storage/innobase/xtrabackup/src/xtrabackup.cc:
```diff
@@ -4676,7 +4875,7 @@ static bool get_meta_path(
   size_t len = strlen(delta_path);
 
-  if (len <= 6 || strcmp(delta_path + len - 6, ".delta")) {
+  if (len <= 6 || strcmp(delta_path + len - 6, EXT_DELTA.c_str())) {
```
possible to avoid 6 ? strlen(EXT_DELTA). Simlar comment to other EXT variants introduced in this code.

aybek replied:
AB

satya-bodapati commented on storage/innobase/xtrabackup/src/xtrabackup.cc:
```diff
@@ -5049,7 +5248,7 @@ static pfs_os_file_t xb_delta_open_matching_space(
 
       xb::info() << "Renaming " << dest_space_name << " to " << tmpname
-                 << ".ibd";
+                 << EXT_IBU;
```
This should be EXT_IBD. Refactoring/cleanup mistake

aybek replied:
AB

satya-bodapati commented on storage/innobase/xtrabackup/src/xtrabackup.cc:
```diff
@@ -6498,6 +6698,78 @@ static void xtrabackup_prepare_func(int argc, char **argv) {
+  if (opt_lock_ddl == LOCK_DDL_REDUCED) {
+    if (!xb_process_datadir(
+            xtrabackup_incremental_dir ? xtrabackup_incremental_dir : ".",
+            ".corrupt", prepare_handle_corrupt_files, NULL)) {
```
new extention for ".CORRUPT". May be introduced a short name ".CRPT" ?

aybek replied:
AB

satya-bodapati commented on storage/innobase/xtrabackup/test/bootstrap.sh:
```diff
@@ -111,7 +111,7 @@ main () {
     case "${TYPE}" in
         innodb80)
-            url="https://dev.mysql.com/get/Downloads/MySQL-8.1"
+            url="https://dev.mysql.com/get/Downloads/MySQL-8.2"
```
8.4

aybek replied:
AB

satya-bodapati commented on storage/innobase/xtrabackup/test/suites/binlog/ib_slave_info.sh:
```diff
@@ -80,7 +80,7 @@ run_cmd_expect_failure $XB_BIN $XB_ARGS --backup --slave-info --no-lock \
 
 vlog "Full backup of the slave server"
-xtrabackup --backup --lock-ddl=false --target-dir=$topdir/backup --slave-info --safe-slave-backup 2>&1 | tee $topdir/pxb.log
+xtrabackup --backup --lock-ddl=OFF --target-dir=$topdir/backup --slave-info --safe-slave-backup 2>&1 | tee $topdir/pxb.log
```
Lets revert these changes from =OFF to =false. It is a valid value.
May be leave some tests with the OFF.

aybek replied:
AB

satya-bodapati commented on storage/innobase/xtrabackup/test/suites/binlog/ib_slave_info.sh:
```diff
@@ -196,7 +202,7 @@ start_server_with_id $slave3_id
 
 vlog "Full backup of the GTID slave server"
-xtrabackup --backup --lock-ddl=false --slave-info --target-dir=$topdir/backup
+xtrabackup --backup --lock-ddl=OFF --slave-info --target-dir=$topdir/backup
```
Lets keep OFF here

aybek replied:
AB

satya-bodapati commented on storage/innobase/xtrabackup/test/suites/binlog/ib_slave_info.sh:
```diff
@@ -165,7 +171,7 @@ setup_slave GTID $slave2_id $master_id
 
 vlog "Full backup of the GTID with AUTO_POSITION slave server"
-xtrabackup --backup --lock-ddl=false --slave-info --target-dir=$topdir/backup
+xtrabackup --backup --lock-ddl=OFF --slave-info --target-dir=$topdir/backup
```
Lets kep OFF here. (Everwhere else, we can use false)

aybek replied:
AB

satya-bodapati commented on storage/innobase/xtrabackup/src/ddl_tracker.cc:
```diff
@@ -0,0 +1,1041 @@
+
+static void delete_force(const std::string &path) {
+  if (access(path.c_str(), R_OK) == 0) {
+    if (my_delete(path.c_str(), MYF(MY_WME))) {
+      exit(EXIT_FAILURE);
```
I think MYF_WME ensure we log the reason for failure on stderr. But lets be safe and write a xb:error() cannot delete.. and error code from strerr(errno)?

aybek replied:
AB

satya-bodapati commented on storage/innobase/xtrabackup/src/ddl_tracker.cc:
```diff
@@ -0,0 +1,1041 @@
+    new_tables[elem.second] = elem.first;
+  }
+
+  if (new_tables.empty()) return;
```
Lets indicate that there are no files are being copied by reduced lock. xb::info() msg please

aybek replied:
AB

Commit:
Addressing the PR review comments

satya-bodapati committed:
Address review comments

satya-bodapati committed:
PXB-3269
1. Move to_string to utils.cc
2. Add checks to ensure the ddl_trackers maps are not updated after
   we reach handle_ddl_operations()

satya-bodapati committed:
PXB-3269 : Address review comments

satya-bodapati commented on storage/innobase/fil/fil0fil.cc:
```diff
@@ -519,8 +519,14 @@ class Tablespace_dirs {
+  @param[in] populate_fil_cache   if true, tabelspace are loaded to cache
+  @param[in] is_prep_handled_ddls if true, xtrabackup uses
+                                  ibd_open_for_recovery instead of
+                                  fil_open_for_xtrabackup()
```
remove

satya-bodapati commented on storage/innobase/fil/fil0fil.cc:
```diff
@@ -1645,6 +1650,11 @@ class Fil_system {
   [[nodiscard]] dberr_t open_for_recovery(space_id_t space_id);
 
+  /** Open a tablespace when LOCK_DDL_REDUCED is on.
+  @param[in]  path    File path
```
Add comment, why we cannot use fil_open_for_xtrabackup() here

satya-bodapati commented on storage/innobase/fil/fil0fil.cc:
```diff
@@ -1700,9 +1710,16 @@ class Fil_system {
+  @param[in] populate_fil_cache   if true, tabelspace are loaded to cache
+  @param[in] is_prep_handled_ddls if true, xtrabackup uses
+                                  ibd_open_for_recovery instead of
+                                  fil_open_for_xtrabackup()
```
remove

satya-bodapati commented on storage/innobase/fil/fil0fil.cc:
```diff
@@ -2704,7 +2720,14 @@ dberr_t Fil_shard::get_file_size(fil_node_t *file, bool read_only_mode) {
   }
 
+  /* space_id mismatch can happen for truncation of tablespaces. Regular and
+   * Undo */
```
formatting

satya-bodapati commented on storage/innobase/fil/fil0fil.cc:
```diff
@@ -11463,6 +11565,14 @@ std::tuple<dberr_t, space_id_t> fil_open_for_xtrabackup(
+  opening renamed table since original table was loaded to cache
+  and copy/rename will be handled by ddl_tracker */
+  if (ddl_tracker && err == DB_TABLESPACE_EXISTS) {
+    ddl_tracker->add_renamed_table(space_id, path);
```
may be rename to add_rename_ibd_scan()

satya-bodapati commented on storage/innobase/xtrabackup/src/ddl_tracker.h:
```diff
@@ -0,0 +1,210 @@
+
+  /** Check if table is in missing list
+  @param[in]  name  tablespace name */
+  bool is_missing_table(const std::string &name);
```
is_missing_after_discovery

satya-bodapati commented on storage/innobase/xtrabackup/src/ddl_tracker.h:
```diff
@@ -0,0 +1,210 @@
+
+  /** Note that a table has been deleted between disovery and file open
+  @param[in]  path  missing table name with path. */
+  void add_missing_table(std::string path);
```
add_missing_after_discovery()

satya-bodapati commented on storage/innobase/xtrabackup/src/ddl_tracker.cc:
```diff
@@ -0,0 +1,1108 @@
+
+void ddl_tracker_t::backup_file_op(uint32_t space_id, mlog_id_t type,
+                                   const byte *buf, ulint len,
+                                   lsn_t start_lsn) {
```
rename to record_lsn, here and everywhere

satya-bodapati commented on storage/innobase/xtrabackup/src/ddl_tracker.cc:
```diff
@@ -0,0 +1,1108 @@
+  while (1) {
+    xb::info() << "Sleeping for 1 second, waiting for the recopy of tables to "
+                  "be finished."
+               << "Expecting count be 0, current count is " << count;
```
space between finished. and Expecting

satya-bodapati commented on storage/innobase/fil/fil0fil.cc:
```diff
@@ -1700,9 +1710,13 @@ class Fil_system {
-  dberr_t scan(bool populate_fil_cache) {
-    return (m_dirs.scan(populate_fil_cache));
+  mapping table
+  @param[in] populate_fil_cache   if true, tabelspace are loaded to cache
```
tablespace*

satya-bodapati commented on storage/innobase/fil/fil0fil.cc:
```diff
@@ -11543,15 +11662,38 @@ void Tablespace_dirs::open_ibd(const Const_iter &start, const Const_iter &end,
-    if (err != DB_SUCCESS && err != DB_INVALID_ENCRYPTION_META) {
-      result = false;
+    /* when LOCK_DDL_REDUCED while processing ddl files on prepare phase
+    we should load data files without first page validation */
```
/* With lock-ddl=reduced, prepare on such backup directories, we open tablespaces before recovery starts. This is required to process to space_id.del or .ren files. Hence, we have to
tolerate encryption errors during first-page validation. Hence, we use a separate opening function of reduced mode */

satya-bodapati commented on storage/innobase/os/os0enc.cc:
```diff
@@ -664,7 +664,9 @@ bool Encryption::decode_encryption_info(space_id_t space_id,
     the table is newly created. For clone encryption information
     should have been already correct. */
-    return (recv_recovery_is_on() ? true : false);
+    bool prep_lock_ddl_reduced =
```
/* With reduced lock ddl, we open just for space_id and filename mapping. Tolerate missing encryption information */

satya-bodapati commented on storage/innobase/xtrabackup/src/xtrabackup.cc:
```diff
@@ -3182,6 +3360,11 @@ static void data_copy_thread_func(data_thread_ctxt_t *ctxt) {
     if (xtrabackup_copy_datafile(node, num)) {
       xb::error() << "failed to copy datafile " << node->name;
       *(ctxt->error) = true;
+    } else {
+      if (ddl_tracker && fsp_is_undo_tablespace(node->space->id)) {
+        ddl_tracker->add_undo_tablespace(node->space->id,
+                                         node->space->files.front().name);
+      }
```
Add a comment that, instead of tracking at the beginning, we will do the tracking here after xtrabackup copies of the undo files.

satya-bodapati commented on storage/innobase/xtrabackup/src/xtrabackup.cc:
```diff
@@ -7445,6 +7719,12 @@ int main(int argc, char **argv) {
   }
 
+  if (opt_page_tracking && opt_lock_ddl == LOCK_DDL_REDUCED) {
+    xb::error() << "--page-tracking and --lock-ddl=REDUCED cannot be enabled "
+                   "together.";
```
fix formatting

satya-bodapati commented on storage/innobase/xtrabackup/src/xtrabackup.h:
```diff
@@ -237,15 +240,22 @@ enum binlog_info_enum {
 void xtrabackup_io_throttling(void);
 bool xb_write_delta_metadata(const char *filename, const xb_delta_info_t *info);
 
+std::string ddl_lock_type_to_str(lock_ddl_type_t type);
+
+lock_ddl_type_t ddl_lock_type_from_str(std::string type);
```
function comments

satya-bodapati commented on storage/innobase/xtrabackup/src/xtrabackup.h:
```diff
@@ -237,15 +240,22 @@ enum binlog_info_enum {
 fil_node_t *datafiles_iter_next(datafiles_iter_t *it);
 void datafiles_iter_free(datafiles_iter_t *it);
 
+bool xtrabackup_copy_datafile_func(fil_node_t *node, uint thread_n,
+                                   const char *dest_name);
+
```
function comment. We are exposing this function now.

satya-bodapati commented on storage/innobase/xtrabackup/src/xtrabackup.h:
```diff
@@ -332,4 +342,6 @@ bool xb_process_datadir(const char *path,   /*!<in: datadir path */
 @param[in]	lsn		lsn to update */
 void update_log_temp_checkpoint(byte *buf, lsn_t lsn);
+
+void xb_scan_for_tablespaces(bool only_undo);
```
function comment. we are exposing this now

satya-bodapati commented on storage/innobase/xtrabackup/test/bootstrap.sh:
```diff
@@ -111,7 +111,7 @@ main () {
     case "${TYPE}" in
         innodb80)
-            url="https://dev.mysql.com/get/Downloads/MySQL-8.1"
+            url="https://dev.mysql.com/get/Downloads/MySQL-8.4"
```
lets fix the PS 8.4 URL too

satya-bodapati committed:
PXB-3269 : Address review comments

satya-bodapati committed:
PXB-3269: Address self review comments

Add ifdef XTRABACKUP for the code introduced in innobase codebase.
(100% not possible though for places that are heavily refactored)

satya-bodapati committed:
PXB-3269 : Add version number to .ren files

satya-bodapati committed:
Add missing function comment

Commit:
PXB-3368 Fix jenkins test failures
Setting handle_ddl_ops variable to false initially

satya-bodapati committed:
PXB-3269 : Fix debug assertion failure on prepare_handle_ren() files.

It is possible that a space_id.ren with content of desired filename,
the destination file name could already exist.

If the source and desitnation to be renamed is same, skip rename.

Commit:
PXB-3368 Fix jenkins test failures

Fix keyring test failures by replacing keyring_file to keyring_component

satya-bodapati committed:
PXB-3370 : Full Backup fails while open file limit is reduced to 1024 from 65536

Problem:
--------
A regression introduced by 6c9aa00 caused extra opening of files.

Fix:
----
remove extra file open

satya-bodapati committed:
PXB-3374 : Space ID  is missing encryption information

With lock-ddl=REDUCED, the following sequence can happen (Not possible
with lock-ddl=ON)

    1. first an IBD with all zero(invalid) encryption is found
    2. so this is added to invalid encryption ids
    3. the same IBD is found again (because of concurrent DDL, they are both found with different names)
    4. this time IBD has proper encryption,  so fil_space_t is created
    5. encryption info from redo is parsed. fil_tablespace_redo_encryption()
    6. because the fil_space_t exists, encryption key is not added to the recovery encryption keys map
    7. later at the end of the backup, we check if we have found a valid encryption key for the invalid encryption space_ids.
    8. we haven’t (remember at step 6, we skipped it)
    9. backup is aborted

Fix:
----
Store keys in recv_sys->keys map even if there exists a tablespace

satya-bodapati committed:
PXB-3381 : Implement check to fail early  if number of open file handles is not same as number of files in datadir

1. Verifies the ulimit -Sn , ulimit -Hn, current --open-files-limit
   parameter and increases the limit if possible, else throws an error
   early.

2. Despite the limit increase, backup may still fail because if there
   are new files that appear, we may need more handles than we first
   calculated.

Commit:
PXB-3380 Backup fails when using external tablespaces and external undo files
Fixed the path for .del/.ren/.new files.
Removed extra scan of external tablespaces during prepare

Commit:
PXB-3380 Extending the tests for external tablespaces

satya-bodapati commented on storage/innobase/xtrabackup/src/ddl_tracker.cc:
```diff
@@ -306,9 +306,16 @@ static void copy_for_reduced(copy_thread_ctxt_t *ctxt) {
+    char dest_name[FN_REFLEN];
+    if (fsp_is_file_per_table(node->space->id, node->space->flags)) {
+      snprintf(dest_name, sizeof(dest_name), "%s%s%s",
+               fil_path_to_space_name(node->name), EXT_IBD.c_str(),
```
Let's use a single place to convert a source filename to a destination filename.


Please move logic fil_path_to_space_name().

May be also also pass node->space->id or the entire node struct. use the space_id to determine if it is undo or not.

mysql> CREATE UNDO TABLESPACE UNDO_25 ADD DATAFILE 'blah_25.ibu';
Query OK, 0 rows affected (0.09 sec)

mysql> CREATE TABLESPACE ts1;
Query OK, 0 rows affected (0.01 sec)

datadir contents:
ls -l var/mysqld.1/data/
total 140612
-rw-rw---- 1 satya satya       56 Oct 10 11:02  auto.cnf
-rw-rw---- 1 satya satya   198343 Oct 10 11:03  binlog.000001
-rw-rw---- 1 satya satya       16 Oct 10 11:02  binlog.index
-rw-rw---- 1 satya satya 16777216 Oct 10 11:04  blah_25.ibu
-rw-rw---- 1 satya satya 16777216 Oct 10 11:03  bloh_25.ibu
-rw-rw---- 1 satya satya   114688 Oct 10 11:03  f3914a36-86ee-11ef-9c27-047bcbcb6b7e.ibd

aybek replied:
done

Commit:
PXB-3380 Addressing review comments

satya-bodapati commented on storage/innobase/xtrabackup/src/ddl_tracker.cc:
```diff
@@ -288,6 +288,36 @@ typedef struct {
+ * @param node File node of a tablespace
+ * @return A string representing the destination name
+ */
+std::string get_dest_name(fil_node_t *node) {
```
we can make parameter const

satya-bodapati commented on storage/innobase/xtrabackup/src/ddl_tracker.cc:
```diff
@@ -288,6 +288,36 @@ typedef struct {
+    db_name = db_name.substr(last_separator_pos + 1);
+  }
+
+  return db_name + '/' + space_name + EXT_NEW;
```
Use Fil_path::SEPARATOR instead of '/'

satya-bodapati commented on storage/innobase/xtrabackup/src/ddl_tracker.cc:
```diff
@@ -306,16 +336,12 @@ static void copy_for_reduced(copy_thread_ctxt_t *ctxt) {
+    char dest_name_buffer[FN_REFLEN];
+    snprintf(dest_name_buffer, FN_REFLEN, "%s", dest_name.c_str());
+
+    if (xtrabackup_copy_datafile_func(node, num, dest_name_buffer)) {
```
simply use dest_name.c_str()

satya-bodapati commented on storage/innobase/xtrabackup/src/ddl_tracker.cc:
```diff
@@ -288,6 +288,36 @@ typedef struct {
+  std::string db_name = path.substr(0, last_separator_pos);
+  std::string space_name = path.substr(last_separator_pos + 1);
+
+  if (!fsp_is_file_per_table(node->space->id, node->space->flags)) {
```
add debug assertion here that space_name shouldn't be empty/(not size 0)

satya-bodapati commented on storage/innobase/xtrabackup/src/ddl_tracker.cc:
```diff
@@ -288,6 +288,36 @@ typedef struct {
+  if (last_separator_pos != std::string::npos) {
+    db_name = db_name.substr(last_separator_pos + 1);
+  }
+
```
also the same for db_name.

add debug assertion here that db_name shouldn't be empty/(not size 0)

Commit:
PXB-3380 Addressing review comments

satya-bodapati committed:
PXB-3387 :  Assertion failure: ddl_tracker.cc:122:!handle_ddl_ops

Problem:
--------
Undo tablespaces are encrypted during the backup. Their key is in
redo but not on the file yet. During recopy phase, xtrabackup was
not able to decrypt the pages and wants to add to list of corrupted
tablespaces. But this is too late as we are at handling DDL operations.

Fix:
----
Read encryption keys of undo tablespaces from redo and use it for
decryption.

satya-bodapati committed:
Followup fix to PXB-3387: fix release build issue

satya-bodapati committed:
PXB-3388 : PXB fails to take backup of general tablespaces created outside data directory when lock-ddl=REDUCED

Problem:
-------
When external general tablespaces are present, with reduced mode,
xtrabackup created the .crpt files in the external dir instead of
the backup dir.

Fix:
---
Always create files within backupdir.

satya-bodapati committed:
PXB-3390: Table Flags Mismatch Causing Assertion Failure in InnoDB Tablespace

Problem:
========
For every IBD file, we open the file twice.
Datafile::open
Fil_shard::get_file_size

Between these two operations, a general tablespace may become
ENCRYPTION Y->N or N->Y

So the tablespace flags can mismatch. This should be tolerated because

Fix:
----
we track this encryption change and recopy them under lock anyway

satya-bodapati committed:
PXB-3269: Jenkins fixes
1. Dumple tablespace key has two parts. Saving the keys during the
   backup and using them during prepare. We should use keys from dump
   file during prepare only
2. Free the memory allocated by fil_path_to_space_name() using a scope
   guard

Commit:
PXB-3393 : PXB fails to take backup of general tablespaces created outside data directory when lock-ddl=REDUCED

    Problem:
    When using backup with lock-ddl=REDUCED and backing up general tablesapce in an external data directory, external directory is assumed as schema of
    the general tablspace. For example:  Delete DDL for `ext_dir/ext_subdir/gen_tbs.ibd` creates `backup/ext_subdir/space_id.del` file, instead of `backup/space_id.del`
    Fix:
    ---
    Check tablespace type and strip the leading path in case of general tablesapces and undo tablespaces

satya-bodapati committed:
probuild changes

satya-bodapati commented:
A newer squashed version will be submitted.
